### DISCUSSION TITLE
vcpkg manifest mode libraries not found in main.cpp

### DISCUSSION THREAD

**Discussion by rockonedege** [friendlyanon/cmake-init#73] [Q&A]
I noticed that in the generated code, there are 2 projects, one for library(lib.h/.cpp) and the other for executable(main.cpp).

when I try to include vcpkg installed library in the main.cpp, it cannot find it. moving the #include <fmt/core.h> from lib.cpp over to main.cpp yieds the same error , while it's perfect fine in the lob.cpp file.

by checking the Cmakelist.txt file, I Couldn't find which line causes this difference.

my questions:
- is it bug or by design that the executable cannot consume the vcpkg libraries, if by design, why?
- What's the trick to achieve this effect?

**Comment by friendlyanon** [comment]
`fmt::fmt` is linked to the library target as a build requirement. If you wish to use fmt elsewhere, you have to link accordingly.

```cmake
find_package(fmt REQUIRED)
             ^ fmt is imported here on line 31
target_link_libraries(proj_lib PRIVATE fmt::fmt)
                      ^        ^
                      |        linked as a build requirement
                      your library target
```

**Reply by rockonedege** [reply]
I might have made you confused.

I used fmt just to verify my obervation that I could #include any vcpkg manifest library in the lib.cpp just fine, but not in the main.cpp, which puzzled me because I this was configured and why.

**Reply by friendlyanon** [reply]
If you want to use fmt in main.cpp, which belongs to the `proj_exe` target, you have to link `fmt::fmt` to it as well.

**Reply by rockonedege** [reply]
the problem is not about linking, it's about include path. and it's not specific to fmt.

the problem is

I could put #include <some.hpp> in lib.cpp and it finds some.hpp and builds.

then I put the same in main.cpp ot says file not found. 

to make sure it's notw some.hpp specific, I Tested with fmt and confirmed it.

I even proved it by checking the generated vcprojx files and the include path was indeed missing from the one for main.cpp.

I couldn't tell which setting contributed to the difference in cmake and vcpkg json files. 

really confused.

**Comment by rockonedege** [comment]
to give more background
```c++
#include <iostream>
#include <string>

#include "lib.hpp"
#include <fmt/core.h> // added to trigger the error

auto main() -> int
{
  auto const lib = library {};
  auto const message = "Hello from " + lib.name + "!";
  std::cout << message << '\n';
  return 0;
}
```
results in this compiling error
```
[build] F:\testme\source\main.cpp(5,1): fatal error C1083: Cannot open include file: 'fmt/core.h': No such file or directory [F:\testme\build\dev-win64\testme_exe.vcxproj]
```

checking the generated vcproj files:

in the generated *_lib.vcproj, it's like this
```xml
    <ClCompile>
      <AdditionalIncludeDirectories>F:\testme\source;%(AdditionalIncludeDirectories)</AdditionalIncludeDirectories>
      <AdditionalOptions>%(AdditionalOptions) /external:I "F:/testme/build/dev-win64/vcpkg_installed/x64-windows-static-md/include" /utf-8 /volatile:iso /Zc:__cplusplus /Zc:externConstexpr /Zc:throwingNew</AdditionalOptions>
```

in the generated *_exe.vcproj, it's like this
```xml
    <ClCompile>
      <AdditionalIncludeDirectories>F:\testme\source;%(AdditionalIncludeDirectories)</AdditionalIncludeDirectories>
      <AdditionalOptions>%(AdditionalOptions) /utf-8 /volatile:iso /Zc:__cplusplus /Zc:externConstexpr /Zc:throwingNew</AdditionalOptions>
```

There's ` /external:I "F:/testme/build/dev-win64/vcpkg_installed/x64-windows-static-md/include"` for the *_lib.vcproj, so 
`#include <fmt/core.h>` is ok in `lib.cpp` but not in `main.cpp`.

Anybody know where this setting is controlled?

**Reply by friendlyanon** [reply]
You need to link to the `fmt::fmt` target to get its include paths:

```cmake
find_package(fmt REQUIRED)
# declare fmt::fmt as a build requirement of testme_exe
target_link_libraries(testme_exe PRIVATE fmt::fmt)
```

**Comment by rockonedege** [comment]
THIS did it.

However I find it counter-intuitive. 

with `find_package(fmt REQUIRED)`, I should be able to find the header to _compile_ it, but fail to _link_ it by missing `target_link_libraries(testme_exe PRIVATE fmt::fmt)`.

Probably I read it too literally and have always got the semantics wrong. Thanks for the help.

**Reply by friendlyanon** [reply]
`find_package(fmt REQUIRED)` finds the `fmt` package.

`target_link_libraries(testme_exe PRIVATE fmt::fmt)` links `fmt::fmt` to your target, getting its properties like defines, libraries, includes, etc.

**Reply by rockonedege** [reply]
Now I Get it.

My previous experience was that， once I have Find_* in the Cmakelists, I could include the headers and have my code compiled with a link error,  then adding the target_link_* removed  the linking error. I used this behavior to find what boost libraries my code depends on,  It was before the manifest mode though
